jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -610,6 +610,15 @@ public boolean onIdleExpired(TimeoutException timeout)
         return false; // We've handle the exception
     }
 
+    @Override
+    public void close()
+    {
+        Runnable task = _httpChannel.onFailure(new EofException());
```

Comment:
I don't like creating an `EofException` for every normal close.   
If you really need this, then create a `_httpChannel.onClose()` method that grabs the state lock, checks the state and only creates the exception if it is actually needed.

Reply:
I added an extra check so the channel is only failed when `_httpChannel.getRequest()` is not null, such as the channel is only failed (and the exception created) only if there is a pending request, which is not the case for a normal close.

That makes the implementation a bit more elegant and takes care of that unneeded exception allocation in the normal case.